rust-runtime/aws-smithy-runtime-api/src/client/retries/classifiers.rs

```diff
@@ -0,0 +1,170 @@
+    /// in case that's important."
+    Error(ErrorKind),
+    /// "The server told us to wait this long before retrying the request."
+    Explicit(Duration),
+    /// "This response should not be retried."
+    DontRetry,
+}
+
+impl fmt::Display for RetryClassifierResult {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Error(kind) => write!(f, "retry ({kind})"),
+            Self::Explicit(duration) => write!(f, "retry after {duration:?}"),
+            Self::DontRetry => write!(f, "don't retry"),
+        }
+    }
+}
+
+/// The priority of a retry classifier. Classifiers with a higher priority will run before
+/// classifiers with a lower priority. Classifiers with equal priorities make no guarantees
+/// about which will run first.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum RetryClassifierPriority {
+    /// The default priority for the [`HttpStatusCodeClassifier`].
+    HttpStatusCodeClassifier,
+    /// The default priority for the [`ModeledAsRetryableClassifier`].
+    ModeledAsRetryableClassifier,
+    /// The default priority for the [`SmithyErrorClassifier`].
+    SmithyErrorClassifier,
+    /// The priority of some other classifier.
+    Other(i8),
+}
+
+impl PartialOrd for RetryClassifierPriority {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(other.as_i8().cmp(&self.as_i8()))
+    }
+}
+
+impl Ord for RetryClassifierPriority {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        other.as_i8().cmp(&self.as_i8())
+    }
+}
+
+impl RetryClassifierPriority {
+    /// Create a new `RetryClassifierPriority` that runs after the given priority.
+    pub fn run_after(other: Self) -> Self {
+        Self::Other(other.as_i8() - 1)
+    }
+
+    /// Create a new `RetryClassifierPriority` that runs before the given priority.
+    pub fn run_before(other: Self) -> Self {
+        Self::Other(other.as_i8() + 1)
+    }
+
+    fn as_i8(&self) -> i8 {
+        match self {
+            Self::HttpStatusCodeClassifier => 0,
+            Self::ModeledAsRetryableClassifier => 10,
+            Self::SmithyErrorClassifier => 20,
+            Self::Other(i) => *i,
+        }
+    }
+}
+
+impl Default for RetryClassifierPriority {
+    fn default() -> Self {
+        Self::Other(0)
+    }
+}
+
+/// Classifies what kind of retry is needed for a given [`InterceptorContext`].
+pub trait ClassifyRetry: Send + Sync + fmt::Debug {
+    /// Run this classifier on the [`InterceptorContext`] to determine if the previous request
+    /// should be retried. Returns a [`RetryClassifierResult`].
+    fn classify_retry(
+        &self,
+        ctx: &InterceptorContext,
+        result_of_preceding_classifier: Option<RetryClassifierResult>,
+    ) -> Option<RetryClassifierResult>;
+
+    /// The name of this retry classifier.
+    ///
+    /// Used for debugging purposes
+    fn name(&self) -> &'static str;
```

Comment:
do we want this only for debugging purposes? Or would we also want to be able to use this (or something like it) to remove classifiers from the list

Reply:
We can change the comment if we decide to implement that feature in the future.